### PR TITLE
documentation: update readme, references and prepare for release

### PR DESCRIPTION
**Pull request by b4ldr**
I have just created a boiler plate README based on extlib, please
 recommend updates

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,50 +1,37 @@
...
+include documentation, and rspec tests ;) but will happily guide you through
+the process, so, yes, please submit that pull request!
 
-Retrieves a DNS CNAME record and returns it as a string.
+Reference documentation is generated using puppet-strings.
+To regenerate it, please run the rake task as follows.
 
-### dns_mx
+```console
+bundle exec rake reference
```
Comment (ekohl) [inline README.md:35]: We want to deprecate this task, in favor of using `strings:generate:reference`. Though I think we should refer to `CONTRIBUTING.md`. If that doesn't cover it, then it's something we should fix instead.

File: README.md [hunk 2]
```diff
@@ -1,50 +1,37 @@
-DNS query functions
-===================
+## Table of Contents
 
-This module contains query functions for DNS for use from Puppet.
+1. [Overview](#overview)
```
Comment (ekohl) [inline README.md:3]: Given the size of the readme, isn't it better to skip the table of contents? GitHub has also introduced it natively: https://github.blog/changelog/2021-04-13-table-of-contents-support-in-markdown-files/ so the practice may be a bit outdated.

File: README.md [hunk 3]
```diff
@@ -1,50 +1,37 @@
...
 
-All functions can take a optional second lambda argument that is called if the lookup returned no values. Then the result of the lambda will be used instead.
+This module provides functions and facts for querying dns records
 
-### dns_lookup
+## Setup
 
-Does a DNS lookup and returns an array of addresses.
+```console
+ % puppet module install puppet-dnsquery
```
Comment (ekohl) [inline README.md:15]: Commonly `$` is used for non-root and `#` for root. I don't see `%` that often for consoles.

File: REFERENCE.md [hunk 4]
```diff
@@ -0,0 +1,746 @@
...
+array will be a IPv4 address.
+
+An optional lambda can be given to return a default value in case the
+lookup fails. The lambda will only be called if the lookup failed.
+
+#### `dns_a(String $record)`
+
+The dns_a function.
+
+Returns: `Any`
```
Comment (ekohl) [inline REFERENCE.md:44]: Perhaps before we merge this we can enhance the functions to specify return types?

File: REFERENCE.md [hunk 5]
```diff
@@ -0,0 +1,746 @@
...
+
+##### `record`
+
+Data type: `String`
+
+
+
+#### `dns_a(String $record, Callable &$block)`
```
Comment (ekohl) [inline REFERENCE.md:52]: The callable is an interesting one. I haven't seen this before. I don't see this in https://github.com/voxpupuli/puppet-dnsquery/blob/master/lib/puppet/functions/dns_a.rb so where does it come from? is a puppet-strings bug or is this a result of the default one?

Comment (alexjfisher) [reply REFERENCE.md:52]: Isn't it this? https://github.com/voxpupuli/puppet-dnsquery/blob/100c8adcf0437c16eb9a0a129c4720c1c506e24f/lib/puppet/functions/dns_a.rb#L15

Comment (alexjfisher) [reply REFERENCE.md:52]: More info here. https://github.com/puppetlabs/puppet-specifications/blob/master/language/func-api.md#blocklambda-support

Comment (ekohl) [reply REFERENCE.md:52]: Those docs suggest you could write a single `dispatch` with `optional_block_param`. That would be easier to follow.

File: REFERENCE.md (hunk 4 above)
Comment (b4ldr) [reply REFERENCE.md:44]: see #126 126

File: REFERENCE.md (hunk 5 above)
Comment (b4ldr) [reply REFERENCE.md:52]: see #126

File: README.md (hunk 1 above)
Comment (b4ldr) [reply README.md]: i updated to use strings:generate:reference.  this repo doesn't have a CONTRIBUTING.md i checked a few other voxpupil projects and couldn't  quickly find one to copy.  can you send a link to one i could add.

File: README.md (hunk 2 above)
Comment (b4ldr) [reply README.md]: done

File: README.md (hunk 3 above)
Comment (b4ldr) [reply README.md]: indeed not many people using the c shell theses days :), updated

File: README.md (hunk 1 above)
Comment (ekohl) [reply README.md]: We store it in `.github`: https://github.com/voxpupuli/puppet-dnsquery/blob/master/.github/CONTRIBUTING.md (comes from https://github.com/voxpupuli/modulesync_config/blob/master/moduleroot/.github/CONTRIBUTING.md.erb)

Comment (b4ldr) [reply README.md]: ahh thanks, added a link

**Comment by b4ldr** [comment]
rebased

File: README.md [hunk 6]
```diff
@@ -1,51 +1,24 @@
...
 
-### dns_lookup
+## Usage
 
-Does a DNS lookup and returns an array of addresses.
+Since `puppet/dnsquery` version 4, all functions are provided under the `dnsquery` namespace.
+Non namespaced versions will be removed in version 5
 
-### dns_a
+All functions are documented in [REFERENCE.md](REFERENCE.md)
```
Comment (ekohl) [inline README.md:16]: I always question these relative links. For example, do they render well on the forge? But we do that more often so I'm not too worried about it.

**Review by ekohl — APPROVED** [review]

File: README.md (hunk 6 above)
Comment (b4ldr) [reply README.md:16]: thanks and ftr i checked they dont work on forge but im not sure what the best thing to do in that case is.  do we want the link to redirect to github, or the rendered reference page on forge?  something to think.  ill merge this as is though

**Pull request merged**